validator/src/admin_rpc_service.rs

```diff
@@ -619,6 +639,51 @@ impl AdminRpc for AdminRpcImpl {
             Ok(())
         })
     }
+
+    fn set_public_tpu_forwards_address(
+        &self,
+        meta: Self::Metadata,
+        public_tpu_forwards_addr: SocketAddr,
+    ) -> Result<()> {
+        debug!("set_public_tpu_forwards_address rpc request received: {public_tpu_forwards_addr}");
+
+        meta.with_post_init(|post_init| {
+            post_init
+                .cluster_info
+                .my_contact_info()
+                .tpu_forwards()
+                .map_err(|err| {
+                    error!(
+                        "The public TPU Forwards address isn't being published. \
+                        The node is likely in repair mode. \
+                        See help for --restricted-repair-only-mode for more information. \
+                        {err}"
+                    );
+                    jsonrpc_core::error::Error::internal_error()
+                })?;
+            post_init
+                .cluster_info
+                .set_tpu_forwards(public_tpu_forwards_addr)
+                .map_err(|err| {
+                    error!("Failed to set public TPU address to {public_tpu_forwards_addr}: {err}");
+                    jsonrpc_core::error::Error::internal_error()
+                })?;
+            warn!(
+                "Public TPU Forwards addresses set to {} (udp) and {} (quic)",
+                post_init
+                    .cluster_info
+                    .my_contact_info()
+                    .tpu_forwards()
+                    .unwrap(),
+                post_init
+                    .cluster_info
+                    .my_contact_info()
+                    .tpu_forwards_quic()
+                    .unwrap(),
```

Comment:
`.my_contact_info()` acquires a read-lock which I _think_ it is not released until the end of this whole `warn!` thing.
So the same thread read-locks the same `RwLock` twice. But `std::sync::RwLock` does not claim to be re-entrant. So this potentially can dead lock. See similar issue in https://github.com/solana-labs/solana/pull/27322

Also, are you sure you want to `unwrap` here for a log?